From ce55bde56f3054d5469fd0ea06aded3906a56c52 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 04 Feb 2022 03:57:56 +0200
Subject: [PATCH] [SW-68803] security : add security init

As a part of boot rom init, security init is added which implements
control of debug ifs and read/store of security keys.

This function can be used in order to append any security related
initializations in boot rom flow.

Change-Id: Ie5b03d9535bc4cee0ec36527ca85f11b8294dcce
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 99da273..80c9ba5 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -458,11 +458,13 @@
 
 __weak void arcboot_rom_disable_spi_gw(void) {}
 __weak void arcboot_rom_wait_for_signal(void) {}
+__weak void arcboot_rom_security_init(void) {}
 
 int arcboot_rom_main(void)
 {
 	arcboot_rom_init();
 	arcboot_rom_fuse();
+	arcboot_rom_security_init();
 	arcboot_rom_disable_spi_gw();
 	arcboot_rom_wait_for_signal();
 	arcboot_rom_in_recovery();
diff --git a/arch/gaudi3/arc/security.c b/arch/gaudi3/arc/security.c
index 5d9bebf..a402333 100644
--- a/arch/gaudi3/arc/security.c
+++ b/arch/gaudi3/arc/security.c
@@ -179,6 +179,16 @@
 	/* TODO: Add additional RR config to lock jtag */
 }
 
+void security_handle_jtag_if(void)
+{
+	if ((arcboot_rom_global.config_fuse &
+		EFUSE_SYSCFG_DISABLE_JTAG_IF_MASK) == 0)
+		security_enable_jtag_if();
+	else
+		security_disable_jtag_if();
+	/* TODO: Add additional RR config to lock jtag */
+}
+
 /* TODO: get newly added key addresses from global specs once it is updated */
 #define mmD0_PSOC_SECURITY_PUBLIC_KEY_3_0		0xC494190
 #define mmD0_PSOC_SECURITY_PUBLIC_KEY_4_0		0xC4941C0
@@ -310,6 +320,19 @@
 		       SECURITY_EFUSE_SYS_CFG_OFFSET);
 }
 
+void arcboot_rom_security_init(void)
+{
+	/* based on efuse, handle jtag enable/disable */
+	security_handle_jtag_if();
+	/* based on efuse, handle twi enable/disable */
+	security_handle_twi();
+	/* store security relevant data in Scratch PAD Sram */
+	security_load_sec_efuse_dword();
+	security_load_revoke_bits();
+	security_load_fw_version();
+	security_load_keys();
+}
+
 int security_efuse_image_validation_en(void)
 {
 	/* check efuse for security validation enable */
